keyvalue/clients/src/main/java/tachyon/client/keyvalue/hadoop/KeyValueOutputCommitter.java

```diff
@@ -34,86 +41,64 @@
-    } catch (TachyonException e) {
-      throw new IOException(e);
+  private List<TachyonURI> getTaskTemporaryStores(JobConf conf) throws IOException {
+    TachyonURI taskOutputURI = KeyValueOutputFormat.getTaskOutputURI(conf);
+    Path outputPath = FileOutputFormat.getOutputPath(conf);
+    FileSystem fs = outputPath.getFileSystem(conf);
+    FileStatus[] subDirs = fs.listStatus(new Path(taskOutputURI.toString()));
+    List<TachyonURI> ret = Lists.newArrayListWithExpectedSize(subDirs.length);
+    for (FileStatus subDir : subDirs) {
+      ret.add(taskOutputURI.join(subDir.getPath().getName()));
     }
-    super.abortJob(context, runState);
+    return ret;
   }
 
   /**
    * {@inheritDoc}
    * <p>
-   * Merges the completed key-value store under the task's temporary output directory to the
-   * key-value store created in {@link #setupJob(JobContext)}.
+   * Merges the completed key-value stores under the task's temporary output directory to the
+   * key-value store created in {@link #setupJob(JobContext)}, then calls
+   * {@link FileOutputCommitter#commitTask(TaskAttemptContext)}.
    */
   @Override
   public void commitTask(TaskAttemptContext context) throws IOException {
-    try {
-      KEY_VALUE_STORES.merge(getTaskAttemptOutputURI(context), getOutputURI(context.getJobConf()));
-    } catch (TachyonException e) {
-      throw new IOException(e);
+    JobConf conf = context.getJobConf();
+    TachyonURI jobOutputURI = KeyValueOutputFormat.getJobOutputURI(conf);
+    for (TachyonURI tempStoreUri : getTaskTemporaryStores(conf)) {
+      try {
+        KEY_VALUE_STORES.merge(tempStoreUri, jobOutputURI);
+      } catch (TachyonException e) {
+        throw new IOException(e);
+      }
     }
+    super.commitTask(context);
   }
 
   /**
    * {@inheritDoc}
    * <p>
-   * Deletes the completed key-value store under the task's temporary output directory, and then
+   * Deletes the completed key-value stores under the task's temporary output directory, and then
    * calls {@link FileOutputCommitter#abortTask(TaskAttemptContext)}.
    */
   @Override
   public void abortTask(TaskAttemptContext context) throws IOException {
-    try {
-      KEY_VALUE_STORES.delete(new TachyonURI(getTaskAttemptPath(context).toString()));
-    } catch (FileDoesNotExistException e) {
-      // The goal of deleting the store is to cleanup directories before aborting the task, since
-      // the key-value store directory does not exist, it meets the goal, nothing needs to be done.
-    } catch (TachyonException e) {
-      throw new IOException(e);
+    for (TachyonURI tempStoreUri : getTaskTemporaryStores(context.getJobConf())) {
+      try {
+        KEY_VALUE_STORES.delete(tempStoreUri);
+      } catch (FileDoesNotExistException e) {
+        // The goal of deleting the store is to cleanup directories before aborting the task, since
+        // the store directory does not exist, it meets the goal, nothing needs to be done.
```

Comment:
after second thought, this exception is unexpected since a temporary store is always created by record writer.